File: web/src/ingest.ts
/**
 * Ingest tokens — `docs/packaging.md` §4.2.
 *
 * The credential that lets a machine somebody else administers send telemetry into this
 * customer's estate. Until it existed the OTLP listener authenticated nobody: whatever could
 * reach the port wrote into its tenant, as any resource it named.
 *
 * # Why this screen is not optional
 *
 * A token that can only be minted with `psql` is a feature nobody has. This product has
 * produced that shape repeatedly — the audit log, user administration, the LLDP walk — and the
 * lesson each time was that a capability is not reachable until there is a screen.
 */

import { request } from "./api";

export interface IngestToken {
  id: string;
  /** What somebody will recognise in six months. */
  label: string;
  created_at: string;
  created_by?: string;
  /** Absent means it does not expire. */
  expires_at?: string;
  revoked_at?: string;
  /** Whether it would authorise a write right now — the server computes it. */
  live: boolean;
}

/** What the server returns once, and never again. */
export interface MintedToken {
  id: string;
  expires_at?: string;
  token: string;
}

export function listTokens(tenant: string): Promise<IngestToken[]> {
  return request<IngestToken[]>("/api/v1/ingest/tokens", { tenant });
}

export function mintToken(
  tenant: string,
  label: string,
  expiresInDays?: number,
): Promise<MintedToken> {
  return request<MintedToken>("/api/v1/ingest/tokens", {
    method: "POST",
    tenant,
    body: { label, expires_in_days: expiresInDays },
  });
}

export function revokeToken(tenant: string, id: string): Promise<void> {
  return request<void>(`/api/v1/ingest/tokens/${id}`, { method: "DELETE", tenant });
}

// ---------------------------------------------------------------------------
// The rules the screen shows, testable without a browser.
// ---------------------------------------------------------------------------

/** Matches `ingest_token`'s `UNIQUE (tenant_id, label)` — a name within one tenant. */
export function labelProblem(label: string, existing: IngestToken[]): string | null {
  const trimmed = label.trim();
  if (trimmed.length === 0) return "A name is required";
  if (trimmed.length > 60) return "At most 60 characters";
  if (existing.some((t) => t.label === trimmed && !t.revoked_at)) {
    return "A live token already uses that name";
  }
  return null;
}

/**
 * What state a token is in, in one word.
 *
 * Three states and not two: *revoked* and *expired* both mean it will not authorise a write,
 * and they mean different things about what happened. An operator scanning for "why did that
 * host stop reporting" needs to tell "somebody revoked this" from "nobody renewed it".
 */
export function tokenState(token: IngestToken, now = new Date()): "live" | "revoked" | "expired" {
  if (token.revoked_at) return "revoked";
  if (token.expires_at && new Date(token.expires_at).getTime() <= now.getTime()) {
    return "expired";
  }
  return "live";
}

/** When it stops working, in words somebody can act on. */
export function expiryNote(token: IngestToken, now = new Date()): string {
  if (!token.expires_at) return "does not expire";
  const ms = new Date(token.expires_at).getTime() - now.getTime();
  if (Number.isNaN(ms)) return "unknown";
  if (ms <= 0) return "expired";
  const days = Math.floor(ms / 86_400_000);
  if (days >= 2) return `expires in ${days} days`;
  const hours = Math.floor(ms / 3_600_000);
  if (hours >= 1) return `expires in ${hours} hour${hours === 1 ? "" : "s"}`;
  return "expires within the hour";
}

/**
 * The command somebody runs on the host they want telemetry from.
 *
 * An OTel Collector exporter block rather than a `curl`, because that is what the thing at the
 * other end actually is — `PLAN` settled that the per-host agent is a distribution of theirs
 * and not an agent of ours, so the useful artefact is their configuration with our endpoint and
 * this token in it.
 *
 * The endpoint is left as a placeholder the operator fills in. The product genuinely does not
 * know it: the listener binds inside a collector on some network, and the address an emitter
 * should use is a routing question about the estate rather than a fact the server holds.
 * Guessing it from the browser's own URL would be wrong in exactly the deployments that matter
 * — a reverse proxy in front, a collector on a different segment.
 */
export function exporterConfig(token: string, endpoint = "<your-otlp-endpoint>"): string {
  return `exporters:
  otlphttp:
    endpoint: http://${endpoint}:4318
    headers:
      Authorization: "Bearer ${token}"

service:
  pipelines:
    metrics: { exporters: [otlphttp] }
    logs:    { exporters: [otlphttp] }`;
}

/**
 * The sentence beside a freshly minted token.
 *
 * Shown once, and the product has no mail transport — the same position
 * `docs/user-administration.md` §4.1 takes for invitations, and for the same reason: an
 * air-gapped installation may never have one.
 */
export const SHOWN_ONCE =
  "Copy this now. Only a hash of it is stored, so it cannot be shown again — if you lose it, \
revoke it and mint another.";

File: crates/uops-api/src/routes/ingest.rs
//! Ingest tokens — `docs/packaging.md` §4.2.
//!
//! ```text
//!   GET    /ingest/tokens        every token this tenant has, without the tokens
//!   POST   /ingest/tokens        mint one; the only moment it exists in the clear
//!   DELETE /ingest/tokens/{id}   revoke one, with immediate effect
//! ```
//!
//! Tenant-scoped and admin, unlike the collector *enrolment* tokens next door which are
//! organization-level. The difference is what each authorises: an enrolment token brings up a
//! collector, which holds database credentials and can serve any tenant assigned to it, so it
//! is an organization's decision. An ingest token authorises writing into **one** tenant, so it
//! is that tenant's own business and an admin of that tenant may mint one.
//!
//! Admin rather than operator, because a token is a credential: it lets a machine write into
//! this customer's estate until somebody revokes it, and acknowledging an alert is a different
//! kind of act.

use axum::Json;
use axum::extract::{Path, State};
use axum::http::StatusCode;
use serde::{Deserialize, Serialize};
use uops_core::{ActorId, Role};

use crate::csrf::CsrfChecked;
use crate::error::{ApiError, ApiResult};
use crate::extract::Caller;
use crate::state::AppState;

#[derive(Debug, Serialize)]
pub struct TokenView {
    pub id: uuid::Uuid,
    /// What somebody will recognise in six months.
    pub label: String,
    pub created_at: String,
    pub created_by: Option<ActorId>,
    /// Absent means it does not expire.
    pub expires_at: Option<String>,
    pub revoked_at: Option<String>,
    /// Whether it would authorise a write right now. Computed here rather than left to the
    /// screen, so "revoked" and "expired" do not each need re-deriving in TypeScript.
    pub live: bool,
}

/// `GET /api/v1/ingest/tokens`
///
/// Includes revoked and expired ones: *what did we hand out* is a question about history, and a
/// list that hid them would answer something else.
pub async fn list(State(state): State<AppState>, caller: Caller) -> ApiResult<Json<Vec<TokenView>>> {
    caller.require(Role::Admin)?;

    let now = chrono::Utc::now();
    let rows = state.store.ingest_tokens(caller.scope()).await?;

    caller.audit().read(
        "ingest.tokens",
        Some(i64::try_from(rows.len()).unwrap_or(i64::MAX)),
    );

    Ok(Json(
        rows.into_iter()
            .map(|t| TokenView {
                // Computed before the move, because `live` reads the fields `label` takes.
                live: t.live(now),
                id: t.id,
                label: t.label,
                created_at: t.created_at.to_rfc3339(),
                created_by: t.created_by,
                expires_at: t.expires_at.map(|d| d.to_rfc3339()),
                revoked_at: t.revoked_at.map(|d| d.to_rfc3339()),
            })
            .collect(),
    ))
}

#[derive(Debug, Deserialize)]
pub struct NewToken {
    pub label: String,
    /// How many days until it expires. Absent means never, which is right for a token living
    /// in a configuration-management repository.
    pub expires_in_days: Option<u32>,
}

#[derive(Debug, Serialize)]
pub struct MintedToken {
    pub id: uuid::Uuid,
    pub expires_at: Option<String>,
    /// **The only moment this exists.** Only its hash is stored, so an operator who loses it
    /// mints another. Returned in the body because the product has no organization-level mail
    /// transport and an air-gapped installation may never have one — the same decision
    /// `docs/user-administration.md` §4.1 made for invitations.
    pub token: String,
}

/// `POST /api/v1/ingest/tokens`
pub async fn mint(
    State(state): State<AppState>,
    caller: Caller,
    _csrf: CsrfChecked,
    Json(body): Json<NewToken>,
) -> ApiResult<(StatusCode, Json<MintedToken>)> {
    caller.require(Role::Admin)?;

    let label = body.label.trim();
    if label.is_empty() {
        return Err(ApiError::BadRequest(
            "a label is required: at revocation time the question is which token this is, and \
             a list of hashes cannot answer it"
                .to_owned(),
        ));
    }

    let expires_at = body
        .expires_in_days
        .filter(|d| *d > 0)
        .map(|d| chrono::Utc::now() + chrono::Duration::days(i64::from(d)));

    let issued = state
        .store
        .issue_ingest_token(caller.scope(), label, expires_at, Some(caller.user_id()))
        .await
        .map_err(|e| match e {
            // The unique violation on (tenant_id, label) is a conflict rather than bad input:
            // nothing about the request was malformed.
            uops_core::Error::Invalid(_) => ApiError::Conflict(
                "a token with that label already exists here. Revoke it, or pick another name"
                    .to_owned(),
            ),
            other => ApiError::Internal(other),
        })?;

    // The label, never the token. The audit log is read by more people than this response is,
    // and a token in it is a working credential sitting in a table.
    caller.audit().wrote(
        "ingest.token.mint",
        issued.id.to_string(),
        None,
        Some(serde_json::json!({ "label": label })),
    );

    Ok((
        StatusCode::CREATED,
        Json(MintedToken {
            id: issued.id,
            expires_at: issued.expires_at.map(|d| d.to_rfc3339()),
            token: issued.token,
        }),
    ))
}

/// `DELETE /api/v1/ingest/tokens/{id}`
///
/// Immediate: the listener asks the store per request and holds no cache, so the next export
/// carrying this token is refused.
pub async fn revoke(
    State(state): State<AppState>,
    caller: Caller,
    Path(id): Path<uuid::Uuid>,
    _csrf: CsrfChecked,
) -> ApiResult<StatusCode> {
    caller.require(Role::Admin)?;

    if !state.store.revoke_ingest_token(caller.scope(), id).await? {
        return Err(ApiError::NotFound);
    }

    caller
        .audit()
        .wrote("ingest.token.revoke", id.to_string(), None, None);
    Ok(StatusCode::NO_CONTENT)
}
